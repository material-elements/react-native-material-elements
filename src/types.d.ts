export * from './components/types';
export * from './types/index';
