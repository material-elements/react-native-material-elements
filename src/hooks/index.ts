export * from './useRestyle';
export * from './useThemedProps';
