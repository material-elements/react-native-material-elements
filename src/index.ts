export * from './components';
export * from './hooks';
export * from './libraries';
export * from './utils';
