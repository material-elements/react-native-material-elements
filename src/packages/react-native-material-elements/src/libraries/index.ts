export * from './themes';
export * from './style';
