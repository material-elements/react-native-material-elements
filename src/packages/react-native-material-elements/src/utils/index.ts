export * from './styleGenerator';
export * from './utils';
export * from './dimensions';
