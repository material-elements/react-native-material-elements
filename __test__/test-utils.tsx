import React from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import { ThemeProvider } from '../src';

export const ThemeWrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return <ThemeProvider>{children}</ThemeProvider>;
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: ThemeWrapper, ...options });

export * from '@testing-library/react-native';
export { customRender as render };
