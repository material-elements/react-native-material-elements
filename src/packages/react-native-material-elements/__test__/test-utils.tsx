import React from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import { ThemeProvider } from '../src';
import { ThemeProviderProps } from '../src/libraries/types';

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  themeProps?: Partial<ThemeProviderProps>;
}

export const ThemeWrapper: React.FC<ThemeProviderProps> = ({ children, ...props }) => {
  return <ThemeProvider {...props}>{children}</ThemeProvider>;
};

export const customRender = (ui: React.ReactElement, options?: CustomRenderOptions) => {
  const { themeProps, ...rest } = options || {};
  return render(ui, {
    wrapper: wrapperProps => <ThemeWrapper {...themeProps}>{wrapperProps.children}</ThemeWrapper>,
    ...rest,
  });
};

export * from '@testing-library/react-native';
export { customRender as render };
