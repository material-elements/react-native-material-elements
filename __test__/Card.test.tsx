import { render as testRenderer, waitFor } from '@testing-library/react-native';
import React from 'react';
import { StyleSheet, View } from 'react-native';
import { Card, CardContent, CardHeader, grey, ThemeProvider } from '../src';
import { render } from './test-utils';

describe('Card Component', () => {
  const mockCardTestId = 'card-test-id';

  const mockRef = React.createRef<View>();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', async () => {
    const { toJSON } = render(<Card />);
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('should forward ref correctly', () => {
    render(<Card ref={mockRef} />);
    expect(mockRef.current).toBeDefined();
    expect(mockRef.current).toBeInstanceOf(View);
  });

  it('should apply the sx styles correctly', () => {
    const mockSx = { bg: 'red' };
    const { getByTestId } = render(<Card sx={mockSx} testID={mockCardTestId} />);

    const card = getByTestId(mockCardTestId);
    const flattenedStyle = StyleSheet.flatten(card.props.style);
    expect(flattenedStyle.backgroundColor).toEqual('red');
  });

  it('should apply the card variant (outlined)', () => {
    const { getByTestId } = render(<Card variation="outlined" testID={mockCardTestId} />);

    const card = getByTestId(mockCardTestId);
    const flattenedStyle = StyleSheet.flatten(card.props.style);
    expect(flattenedStyle.borderWidth).toEqual(0.5);
    expect(flattenedStyle.borderColor).toEqual(grey[500]);
  });

  it('should apply the root style', () => {
    const { getByTestId } = testRenderer(
      <ThemeProvider components={{ cardProps: { style: { backgroundColor: 'red' } } }}>
        <Card variation="outlined" testID={mockCardTestId} />
      </ThemeProvider>,
    );
    const card = getByTestId(mockCardTestId);
    const flattenedStyle = StyleSheet.flatten(card.props.style);
    expect(flattenedStyle.backgroundColor).toEqual('red');
  });

  it('should combine the root style and component style', () => {
    const { getByTestId } = testRenderer(
      <ThemeProvider components={{ cardProps: { style: { backgroundColor: 'red' } } }}>
        <Card variation="outlined" testID={mockCardTestId} style={{ borderWidth: 2 }} />
      </ThemeProvider>,
    );
    const card = getByTestId(mockCardTestId);
    const flattenedStyle = StyleSheet.flatten(card.props.style);
    expect(flattenedStyle.backgroundColor).toEqual('red');
    expect(flattenedStyle.borderWidth).toEqual(2);
  });
});

describe('CardHeader Component', () => {
  const mockCardHeaderTestId = 'card-header-test-id';

  const mockRef = React.createRef<View>();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', async () => {
    const { toJSON } = render(<CardHeader />);
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('should forward ref correctly', () => {
    render(<CardHeader ref={mockRef} />);
    expect(mockRef.current).toBeDefined();
    expect(mockRef.current).toBeInstanceOf(View);
  });

  it('should apply the sx styles correctly', () => {
    const mockSx = { bg: 'red' };
    const { getByTestId } = render(<CardHeader sx={mockSx} testID={mockCardHeaderTestId} />);

    const cardHeader = getByTestId(mockCardHeaderTestId);
    const flattenedStyle = StyleSheet.flatten(cardHeader.props.style);
    expect(flattenedStyle.backgroundColor).toEqual('red');
  });

  it('should apply the root style', () => {
    const { getByTestId } = testRenderer(
      <ThemeProvider components={{ cardHeaderProps: { style: { backgroundColor: 'red' } } }}>
        <CardHeader testID={mockCardHeaderTestId} />
      </ThemeProvider>,
    );
    const cardHeader = getByTestId(mockCardHeaderTestId);
    const flattenedStyle = StyleSheet.flatten(cardHeader.props.style);
    expect(flattenedStyle.backgroundColor).toEqual('red');
  });

  it('should combine the root style and component style', () => {
    const { getByTestId } = testRenderer(
      <ThemeProvider components={{ cardHeaderProps: { style: { backgroundColor: 'red' } } }}>
        <CardHeader testID={mockCardHeaderTestId} style={{ borderWidth: 2 }} />
      </ThemeProvider>,
    );
    const cardHeader = getByTestId(mockCardHeaderTestId);
    const flattenedStyle = StyleSheet.flatten(cardHeader.props.style);
    expect(flattenedStyle.backgroundColor).toEqual('red');
    expect(flattenedStyle.borderWidth).toEqual(2);
  });
});

describe('CardContent Component', () => {
  const mockCardContentTestId = 'card-content-test-td';

  const mockRef = React.createRef<View>();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', async () => {
    const { toJSON } = render(<CardContent />);
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('should forward ref correctly', () => {
    render(<CardContent ref={mockRef} />);
    expect(mockRef.current).toBeDefined();
    expect(mockRef.current).toBeInstanceOf(View);
  });

  it('should apply the sx styles correctly', () => {
    const mockSx = { bg: 'red' };
    const { getByTestId } = render(<CardContent sx={mockSx} testID={mockCardContentTestId} />);

    const cardHeader = getByTestId(mockCardContentTestId);
    const flattenedStyle = StyleSheet.flatten(cardHeader.props.style);
    expect(flattenedStyle.backgroundColor).toEqual('red');
  });
});
