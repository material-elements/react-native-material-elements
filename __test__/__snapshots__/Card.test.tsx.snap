// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component should render correctly 1`] = `
<View
  style={
    [
      {},
      {},
      [
        undefined,
        undefined,
        {},
        undefined,
      ],
    ]
  }
/>
`;

exports[`CardContent Component should render correctly 1`] = `
<View
  style={
    [
      {},
      {},
      {
        "padding": 10,
      },
    ]
  }
/>
`;

exports[`CardHeader Component should render correctly 1`] = `
<View
  style={
    [
      {},
      {},
      [
        undefined,
        {},
        undefined,
      ],
    ]
  }
/>
`;
